compiler/rustc_const_eval/src/const_eval/eval_queries.rs

```diff
@@ -294,6 +294,29 @@ pub fn eval_to_const_value_raw_provider<'tcx>(
                 )
             },
         );
+    } else if let ty::InstanceKind::Item(def_id) = key.value.instance.def
+        && matches!(tcx.def_kind(def_id), DefKind::Const | DefKind::AssocConst)
+    {
+        let ct = tcx.const_of_item(def_id).instantiate(tcx, key.value.instance.args);
+        match ct.kind() {
+            ty::ConstKind::Unevaluated(_) => {
+                return Err(ErrorHandled::TooGeneric(DUMMY_SP));
+            }
+            ty::ConstKind::Value(cv) => return Ok(tcx.valtree_to_const_val(cv)),
+            ty::ConstKind::Error(guar) => {
+                return Err(ErrorHandled::Reported(
+                    ReportedErrorInfo::const_eval_error(guar),
+                    DUMMY_SP,
+                ));
+            }
+            ty::ConstKind::Expr(_) => return Err(ErrorHandled::TooGeneric(DUMMY_SP)),
+            ty::ConstKind::Param(_) | ty::ConstKind::Placeholder(_) => {
+                return Err(ErrorHandled::TooGeneric(DUMMY_SP));
+            }
+            ty::ConstKind::Infer(_) | ty::ConstKind::Bound(..) => {
+                bug!("unexpected constant {ct:?}")
+            }
+        }
```

Comment:
I think this is in the slightly wrong place after all. `eval_to_allocation_raw_provider` seems to be the common denominator of all the CTFE entry points so this is currently missing some stuff. If you move it though I think `ConstKind::Value` is going to become a complete pain in the ass to support so we should possibly just *not* support that.

We should change `lower_anon_const` to not special case lowering `{ 1 }` to `Value(0x1)` when the anon const is the rhs of a const item. Then we'll never encounter a `ConstKind::Value` here and dont have to worry about trying to convert it to an allocation or whatever.